### PR TITLE
FSEvents: multiple watchers

### PR DESCRIPTION
**Pull request by rgrinberg**
Allow to create multiple fsevents watchers and receive all the events in one thread. This is a pre-req for watching external directories.

### CODE REVIEW

File: src/dune_file_watcher/dune_file_watcher.ml [hunk 1]
```diff
@@ -217,7 +300,11 @@ module Buffer = struct
 end
 
 let special_file_for_inotify_sync =
-  let path = lazy (Path.Build.relative Path.Build.root "dune-inotify-sync") in
+  let path =
+    lazy
+      (let dir = Path.Build.relative Path.Build.root ".sync" in
+       Path.Build.relative dir "dune-inotify-sync")
```
Comment (jeremiedimino) [inline src/dune_file_watcher/dune_file_watcher.ml:306]: Why this change?

Comment (rgrinberg) [reply src/dune_file_watcher/dune_file_watcher.ml:306]: It is needed to use fsevents more efficiently. In fsevents, all watches are recursive. So to watch for the creation of _build/dune-inotify-sync, we'd need to watch all of _build recursively. That seems a bit wasteful, hence this workaround. I considered other work arounds as well:

* Polling for this file.
* Using kevent to wait for the creationg - kevent allows to wait for a while to be created.

But ultimately, this file is only needed for tests so I just went with what was more expedient and moved it where the watch wouldn't be a problem.

Comment (jeremiedimino) [reply src/dune_file_watcher/dune_file_watcher.ml:306]: Ok, it's not super important so I won't bother you more with this. However, this `inotify-sync` trick is only used because inotify doesn't have a "flush" operation that allows to wait for all pending events to be reported. I thought `fsevents` had one and so we didn't needed the inotify-sync trick with fsevents.

As a side note, if this is now used for both inotify and fsevents, it should probably be called something more general than "inotify-sync".

**Review by jeremiedimino — APPROVED** [review]
Looks good!

File: src/dune_file_watcher/dune_file_watcher.ml (hunk 1 above)
Comment (rgrinberg) [reply src/dune_file_watcher/dune_file_watcher.ml:306]: fsevents has a flush but it's not quite the same thing. I believe it can be used to flush events that are internally buffered, and not to wait for all pending events to arrive. At least this is how I understood after experimenting.

You're right, I should rename this file.
